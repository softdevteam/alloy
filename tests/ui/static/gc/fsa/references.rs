#![feature(gc)]
#![feature(negative_impls)]
#![allow(dead_code)]
#![allow(unused_variables)]
include!{"./auxiliary/types.rs"}

impl<'a> Drop for HasRef<'a> {
    fn drop(&mut self) {
        use_val(self.a); // should fail
        use_val(self.b); // should pass
        use_val(self.c[0]); // should fail

        let a = self.a; // should fail
        let b = self.b; // should pass
        let c = self.c;
        use_val(c[1]); // should fail

        // should pass, as not a field projection
        let c = &1;
        use_val(c);
    }
}

fn main() {
    std::gc::Gc::new(HasRef::default());
    //~^     ERROR: `HasRef::default()` has a drop method which cannot be safely finalized.
    //~^^    ERROR: `HasRef::default()` has a drop method which cannot be safely finalized.
    //~^^^   ERROR: `HasRef::default()` has a drop method which cannot be safely finalized.
    //~^^^^  ERROR: `HasRef::default()` has a drop method which cannot be safely finalized.
    //~^^^^^ ERROR: `HasRef::default()` has a drop method which cannot be safely finalized.
}
