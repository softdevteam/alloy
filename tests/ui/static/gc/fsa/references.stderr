error: `HasRef::default()` has a drop method which cannot be safely finalized.
  --> $DIR/references.rs:25:22
   |
LL |         use_val(self.a); // should fail
   |                 ------
   |                 |
   |                 caused by the expression here in `fn drop(&mut)` because
   |                 it is a reference (&u64) which is not safe to use in a finalizer.
...
LL |     std::gc::Gc::new(HasRef::default());
   |                      ^^^^^^^^^^^^^^^^^
   |
   = help: `Gc` may run finalizers after the valid lifetime of this reference.

error: `HasRef::default()` has a drop method which cannot be safely finalized.
  --> $DIR/references.rs:25:22
   |
LL |         use_val(self.c[0]); // should fail
   |                 ---------
   |                 |
   |                 caused by the expression in `fn drop(&mut)` here because
   |                 it uses a type which is not safe to use in a finalizer.
...
LL |     std::gc::Gc::new(HasRef::default());
   |     -----------------^^^^^^^^^^^^^^^^^- `Gc::new` requires that [&u64; 2] implements the `FinalizeSafe` trait.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values whose types implement `FinalizerSafe`.

error: `HasRef::default()` has a drop method which cannot be safely finalized.
  --> $DIR/references.rs:25:22
   |
LL |         let a = self.a; // should fail
   |                 ------
   |                 |
   |                 caused by the expression here in `fn drop(&mut)` because
   |                 it is a reference (&u64) which is not safe to use in a finalizer.
...
LL |     std::gc::Gc::new(HasRef::default());
   |                      ^^^^^^^^^^^^^^^^^
   |
   = help: `Gc` may run finalizers after the valid lifetime of this reference.

error: `HasRef::default()` has a drop method which cannot be safely finalized.
  --> $DIR/references.rs:25:22
   |
LL |         let c = self.c;
   |                 ------
   |                 |
   |                 caused by the expression in `fn drop(&mut)` here because
   |                 it uses a type which is not safe to use in a finalizer.
...
LL |     std::gc::Gc::new(HasRef::default());
   |     -----------------^^^^^^^^^^^^^^^^^- `Gc::new` requires that [&u64; 2] implements the `FinalizeSafe` trait.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values whose types implement `FinalizerSafe`.

error: `HasRef::default()` has a drop method which cannot be safely finalized.
  --> $DIR/references.rs:25:22
   |
LL |         use_val(c[1]); // should fail
   |                 ----
   |                 |
   |                 caused by the expression here in `fn drop(&mut)` because
   |                 it is a reference (&u64) which is not safe to use in a finalizer.
...
LL |     std::gc::Gc::new(HasRef::default());
   |                      ^^^^^^^^^^^^^^^^^
   |
   = help: `Gc` may run finalizers after the valid lifetime of this reference.

error: aborting due to 5 previous errors

